scheduling/taskmetadata.go

```diff
@@ -0,0 +1,533 @@
+		err.AddDetail("run_id", string(runID))
+		err.AddDetail("end_time", time.String())
+		return err
+	}
+
+	metadata, e := m.GetRunByID(taskID, runID)
+	if e != nil {
+		e.AddDetail("task_id", string(taskID))
+		e.AddDetail("run_id", string(runID))
+		return e
+	}
+
+	updateEndTime := func(runMetadata string) (string, fferr.GRPCError) {
+		metadata := TaskRunMetadata{}
+		err := metadata.Unmarshal([]byte(runMetadata))
+		if err != nil {
+			err.AddDetail("task_id", string(taskID))
+			err.AddDetail("run_id", string(runID))
+			return "", err
+		}
+
+		if metadata.StartTime.After(time) {
+			err := fferr.NewInvalidArgumentError(fmt.Errorf("end time cannot be before start time"))
+			err.AddDetail("task_id", string(taskID))
+			err.AddDetail("run_id", string(runID))
+			err.AddDetail("start_time", metadata.StartTime.String())
+			err.AddDetail("end_time", time.String())
+			return "", err
+		}
+
+		metadata.EndTime = time
+		serializedMetadata, e := metadata.Marshal()
+		if e != nil {
+			return "", e
+		}
+
+		return string(serializedMetadata), nil
+	}
+
+	taskRunMetadataKey := TaskRunMetadataKey{taskID: taskID, runID: metadata.ID, date: metadata.StartTime}
+	e = m.storage.Update(taskRunMetadataKey.String(), updateEndTime)
+	return e
+}
+
+func (m *TaskMetadataManager) AppendRunLog(runID TaskRunID, taskID TaskID, log string) fferr.GRPCError {
+	if runID <= 0 {
+		err := fferr.NewInvalidArgumentError(fmt.Errorf("run id has to be greater than 0"))
+		err.AddDetail("run_id", string(runID))
+		err.AddDetail("task_id", string(taskID))
+		return err
+	}
+	if taskID <= 0 {
+		err := fferr.NewInvalidArgumentError(fmt.Errorf("task id has to be greater than 0"))
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return err
+	}
+	if log == "" {
+		err := fferr.NewInvalidArgumentError(fmt.Errorf("log cannot be empty"))
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return err
+	}
+
+	metadata, e := m.GetRunByID(taskID, runID)
+	if e != nil {
+		return e
+	}
+
+	updateLog := func(runMetadata string) (string, fferr.GRPCError) {
+		metadata := TaskRunMetadata{}
+		err := metadata.Unmarshal([]byte(runMetadata))
+		if err != nil {
+			return "", err
+		}
+
+		metadata.Logs = append(metadata.Logs, log)
+
+		serializedMetadata, e := metadata.Marshal()
+		if e != nil {
+			return "", err
+		}
+
+		return string(serializedMetadata), nil
+	}
+
+	taskRunMetadataKey := TaskRunMetadataKey{taskID: taskID, runID: metadata.ID, date: metadata.StartTime}
+	e = m.storage.Update(taskRunMetadataKey.String(), updateLog)
+	return e
+}
+
+// Finds the highest increment in a list of strings formatted like "/tasks/metadata/task_id=0"
+func getLatestId(taskMetadata map[string]string) (int, fferr.GRPCError) {
+	highestTaskId := -1
+
+	// Regular expression pattern to match "task_id=<number>"
+	pattern := regexp.MustCompile(`/tasks/metadata/task_id=(\d+)`)
+
+	for path, _ := range taskMetadata {
```

Comment:
you can shorthand this if you only need the path

`for path := range taskMetadata`